### DISCUSSION TITLE
Share capture between hurl runs? Setup with multiple files?

### DISCUSSION THREAD

**Discussion by oschrenk** [Orange-OpenSource/hurl#3148] [General] [answered]
TL;DR: Can I write a captured cookie to a file? (and then use that file for a different hurl call)? How can I chain captures betwen multiple files eg. `hurl login.hurl dotest.hurl`?

---

I'm trying to use hurl for building an integration test suite. I like to separate login from other test cases.

How can I separate out the login procedure and share it subsequently with other executions?

I tried two approaches:

1. Using `--output` - but that writes the body of the login response to a file and not the captured variable.
2. Using `hurl login.hurl booking.hurl` but captures from the first file are not being transferred to the second argument.

How would I be able to use a "multi-file" approach? How would I be able to share a login between different testing flows? Or is `hurl` designed to be working only with a single file?

---

Background: 

Idea is something like:
- use `hurl login.hurl --output-capture variables.txt` to write the cookie value to a file
- feed that variable file to the next call like `hurl --variable-file variables.txt test1.hurl | jq . > test1-prett1.json`
- repeat test calls

**Comment by oschrenk** [comment]
Some progress

I found this [issue thread](https://github.com/Orange-OpenSource/hurl/issues/317). It offered these ideas:

- simple cat `cat login.hurl action.hurl | hurl`
- using preprocessor see https://gist.github.com/nikeee/748301705d3ed53b6b8776fba8748a9b that enabled  import statements 

That alleviates some of the issues. I realized though that I might need to create loops and templating, so I need some more power then `cat` or simple imports.


In the meantime I also found out that `--verbose` does output the capture. It's a bit ugly but one can do something like

```shell
hurl --verbose login.hurl 2>&1 | grep -A 1 "\* Captures" | tail -n +2 | awk -F awk -F '[ :]' '{print "export HURL_"$2"="$4}'
export HURL_session-id=cafebabecafebabe

```

(and increase `-A 1` to the number of captures)

**Comment by jcamiel** [comment, marked as answer]
Hi @oschrenk 

There are other possibilities to share data between tests:

- use [`--cookie`](https://hurl.dev/docs/manual.html#cookie) and [`--cookie-jar`](https://hurl.dev/docs/manual.html#cookie-jar): these options allow you to save and restore cookies so you can "simulate" a session based test across filed
- use [`--json`](https://hurl.dev/docs/manual.html#json): this option exports a run result as structured data; you can get a JSON result with headers, captures , cookies etc...

Things to note:

- cookies are automatically managed when running one Hurl file so, as you have noticed, the best use case for Hurl is when tests are self contained, isolated (like unit tests for instance)
- following the unit tests comparison, with `--test`, Hurl files will be run in parallel to speed up the execution. It you rely on execution order, tou can :
    - split up your tests 
    ```
     $ hurl setup.hurl
     $ hurl --test integration/*.hurl
     $ hurl teardown.hurl
    ```
    - run tests in sequential with `--jobs 1` option: `hurl --test --jobs 1 *.hurl'

**Reply by oschrenk** [reply]
Ah, these built in options are what I was looking for! Thank you!
